sw/device/lib/dif/dif_gpio.c

```diff
@@ -0,0 +1,374 @@
+ * @param val Value to write.
+ */
+static void gpio_masked_bit_write(const dif_gpio_t *gpio, uint32_t lower_offset,
+                                  uint32_t upper_offset, uint32_t index,
+                                  bool val) {
+  // Write to lower_offset if the bit is in the lower half, write to
+  // upper_offset otherwise.
+  const ptrdiff_t offset = (index < 16) ? lower_offset : upper_offset;
+  // Since masked access writes to half of a register, index mod 16 gives the
+  // index of the bit in the half-word mask.
+  const uint32_t mask = index_to_mask(index % 16);
+  mmio_region_write32(gpio->base_addr, offset,
+                      (mask << 16) | (val ? mask : 0u));
+}
+
+dif_gpio_result_t dif_gpio_init(const dif_gpio_config_t *config,
+                                dif_gpio_t *gpio) {
+  if (config == NULL || gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  // Save internal state in the given |dif_gpio_t| instance.
+  gpio->base_addr = config->base_addr;
+  // Reset the GPIO device at the given |base_addr|.
+  dif_gpio_reset(gpio);
+
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_reset(const dif_gpio_t *gpio) {
+  if (gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  mmio_region_write32(gpio->base_addr, GPIO_INTR_ENABLE_REG_OFFSET, 0);
+  mmio_region_write32(gpio->base_addr, GPIO_DIRECT_OE_REG_OFFSET, 0);
+  mmio_region_write32(gpio->base_addr, GPIO_DIRECT_OUT_REG_OFFSET, 0);
+  mmio_region_write32(gpio->base_addr, GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, 0);
+  mmio_region_write32(gpio->base_addr, GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, 0);
+  mmio_region_write32(gpio->base_addr, GPIO_INTR_CTRL_EN_LVLHIGH_REG_OFFSET, 0);
+  mmio_region_write32(gpio->base_addr, GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, 0);
+  mmio_region_write32(gpio->base_addr, GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0);
+
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_all_read(const dif_gpio_t *gpio,
+                                    uint32_t *pin_values) {
+  if (gpio == NULL || pin_values == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  *pin_values = mmio_region_read32(gpio->base_addr, GPIO_DATA_IN_REG_OFFSET);
+
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_pin_read(const dif_gpio_t *gpio, uint32_t index,
+                                    bool *pin_value) {
+  if (gpio == NULL || pin_value == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  *pin_value =
+      mmio_region_get_bit32(gpio->base_addr, GPIO_DATA_IN_REG_OFFSET, index);
+
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_all_write(const dif_gpio_t *gpio, uint32_t val) {
+  if (gpio == NULL) {
+    return kDifGpioResultInvalidArgument;
+  }
+
+  mmio_region_write32(gpio->base_addr, GPIO_DIRECT_OUT_REG_OFFSET, val);
+
+  return kDifGpioResultOK;
+}
+
+dif_gpio_result_t dif_gpio_pin_write(const dif_gpio_t *gpio, uint32_t index,
+                                     bool val) {
```

Comment:
What is the consensus on using `bool` for `set/unset, enable/disable`, etc? I use enums for that, as suggested by @mcy . I am not really have a strong opinion on either way, but it would be nice to pick one and stick with it.

Reply:
I agree, but I'm not sure if we have a strict rule for this. I find `bool` natural for a GPIO pin but we can always revisit this if needed.